File: 2816_double_a_number_represented_as_a_linked_list/doubleIt_two_pointers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

        
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current, prev = head, None

        while current:
            val = current.val * 2

            if val < 10:
                current.val = val
            elif prev:
                current.val = val % 10
                prev.val += 1
            else:
                head = ListNode(1, current)
                current.val = val % 10

            prev = current
            current = current.next

        return head

File: 2816_double_a_number_represented_as_a_linked_list/doubleIt_two_pointers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode doubleIt(ListNode head) {
        ListNode current = head;
        ListNode prev = null;

        while (current != null) {
            int value = current.val * 2;

            if (value < 10) {
            current.val = value;
            } else if (prev != null) {
                current.val = value % 10;
                prev.val++;
            } else {
                head = new ListNode(1, current);
                current.val = value % 10;
            }

            prev = current;
            current = current.next;
        }

        return head;
    }
}